### ISSUE TITLE
Issue with parser on newer versions of Electron

### ISSUE THREAD

**Issue by DreamitMauri** [MatthieuLemoine/push-receiver#23]
Using this package on Electron 3x or 4x results in the app crashing with the following error message:

```
npm ERR! code ELIFECYCLE
npm ERR! errno 3221225477
npm ERR! kodinportti@1.0.8 electron:serve: `wait-on http-get://localhost:4200/ && npm run electron:serve-tsc && electron . --serve`
npm ERR! Exit status 3221225477
npm ERR!
npm ERR! Failed at the app@1.0.8 electron:serve script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\user\AppData\Roaming\npm-cache\_logs\2019-01-10T13_12_27_228Z-debug.log
ERROR: "electron:serve" exited with 3221225477.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! app@1.0.8 start: `npm run postinstall:electron && npm-run-all -p ng:serve electron:serve`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the app@1.0.8 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\user\AppData\Roaming\npm-cache\_logs\2019-01-10T13_12_27_396Z-debug.log
```

After digging around the source code I've isolated the problem to somewhere in the Parser.

Enabling debugging in the parser gives these messages:

```
Got data: 1
waitForData state: 0
Socket read finished prematurely. Waiting for 2 more bytes
Got data: 88
waitForData state: 0
Processing MCS data: state == 0
VERSION IS 41
RECEIVED PROTO OF TYPE 3
Proto size: 86
waitForData state: 3
Processing MCS data: state == 3
GCM Handshake complete.
waitForData state: 1
Socket read finished prematurely. Waiting for 2 more bytes
Got data: 105
waitForData state: 1
Processing MCS data: state == 1
RECEIVED PROTO OF TYPE 7
Proto size: 103
waitForData state: 3
Processing MCS data: state == 3
waitForData state: 1
Socket read finished prematurely. Waiting for 2 more bytes
Got data: 514
waitForData state: 1
Processing MCS data: state == 1
RECEIVED PROTO OF TYPE 8
Proto size: 534
waitForData state: 3
Socket read finished prematurely. Waiting for 23 more bytes
Got data: 23
waitForData state: 3
Processing MCS data: state == 3
```

**Comment by nshelia** [comment]
Just checked with debug mode and found issue on [this line](https://github.com/MatthieuLemoine/push-receiver/blob/master/src/utils/decrypt/index.js#L13), which is i think related to Electron ([Github Issue](https://github.com/electron/electron/issues/16477))

**Comment by MatthieuLemoine** [comment]
Which versions of Node are used in Electron 3.x and 4.x ?

**Comment by nshelia** [comment]
https://github.com/electron/node

Node 10.11 is used in Electron 4.0.x

**Comment by jamesb3ll** [comment]
Like @nshelia and @chevonc have [referenced](https://github.com/electron/electron/issues/16477), calling `crypto.createECDH(...).setPrivateKey` is making Electron crash. I replaced `cryto.createECDH` with [`create-ecdh`](https://github.com/crypto-browserify/createECDH/blob/master/browser.js) module in [`decrypt.js`](https://github.com/MatthieuLemoine/push-receiver/blob/master/src/utils/decrypt/index.js#L12) and it works as expected without crashes. Let's hope Electron fixes this soon..

**Comment by chevonc** [comment]
Ahh, good to know @jamesb3ll. it seems that there's some traction on the Electrong PR, but may try out this workaround for now if they don't merge in a couple of days.

https://github.com/electron/electron/pull/16822

**Comment by MatthieuLemoine** [comment]
It seems that the fix was shipped in Electron [4.0.5](https://github.com/electron/electron/releases/tag/v4.0.5) and Electron [5.0.0-beta.3](https://github.com/electron/electron/releases/tag/v5.0.0-beta.3). 

@DreamitMauri @chevonc @nshelia @jamesb3ll could you try if upgrading Electron fixes this issue ?

**Comment by jamesb3ll** [comment]
Hey @MatthieuLemoine, I don't think that Electron [#16912](https://github.com/electron/electron/pull/16912) and [#16909](https://github.com/electron/electron/pull/16909) that was fixed in Electron 4.0.5 and 5.0.0-beta.3 is related to this issue. As pointed out [here](https://github.com/electron/electron/issues/16477#issuecomment-462572613) by [@nornagon](https://github.com/nornagon) that works on Electron.

I think solving [#16477](https://github.com/electron/electron/issues/16477) would fix the problem, it seems related to the patch of node/v8's buffers compatibility in Electron v4+. I'm not sure quite how to solve it and waiting for a reply from [@codebytere](https://github.com/codebytere) on that thread.

**Comment by chevonc** [comment]
Fix for this was merged if anyone else arrives here: https://github.com/electron/electron/releases/tag/v4.1.0

**Comment by DreamitMauri** [comment]
Confirmed. Fixed in Electron 4.1.0.
